model/qti/parser/IncludedElementIdsExtractor.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/*
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoQtiItem\model\qti\parser;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\tao\model\media\TaoMediaResolver;
+use oat\taoQtiItem\model\qti\Item;
+use tao_helpers_Uri;
+
+class IncludedElementIdsExtractor extends ConfigurableService
+{
+    public const OPTION_MEDIA_RESOLVER = 'mediaResolver';
+
+    /** @var string[] */
+    private $ids;
+
+    public function extract(Item $qtiItem): array
+    {
+        $this->ids = [];
+
+        $this->incrementIds($this->normalizeItemBody($qtiItem));
+
+        return array_unique($this->ids);
+    }
+
+    private function normalizeItemBody(Item $qtiItem): array
+    {
+        return json_decode(json_encode($qtiItem->toArray()['body']), true);
+    }
+
+    private function incrementIds(array $body): void
+    {
+        foreach ($body as $key => $value) {
+            $this->incrementIdsByKey($key, $value);
+        }
+    }
+
+    private function incrementIdsByKey(string $key, $value): void
+    {
+        if (!is_array($value)) {
+            return;
+        }
+
+        if ('elements' !== $key) {
+            $this->incrementIds($value);
+
+            return;
+        }
+
+        foreach ($value as $element) {
+            $this->incrementIdsByElement((array)$element);
+        }
+    }
+
+    private function incrementIdsByElement(array $element): void
```

Comment:
Are u going accross all item elements or only accross media elements?

Reply:
Across all the elements that contain `elements` key. It was needed, cause some cases we have sharedStimulus nested to other elements. What do you suggest?

Reply:
need to investigate, to use Item parser

Reply:
Followed your approach! Thanks @siwane . Updated!